### DISCUSSION TITLE
Variable Layer exposure timing based on pct of overhang pixels

### DISCUSSION THREAD

**Discussion by jimb15122** [sn4k3/UVtools#526] [Ideas]
Is it possible to implement the concept of using a min and max exposure time, with the actual time for the layer extrapolated between min and max based on the ratio of overhang pixels in the layer.  e.g. if the layer has no overhang pixels, use the min time, 25% or more, use the max time, else map( 0%, 25%, min, max).

This could be useful for reducing print times, and reducing chances of failed prints.

**Comment by sn4k3** [comment]
It's possible but you will have to write an [script](https://github.com/sn4k3/UVtools/tree/master/UVtools.ScriptSample) to meet your criteria.
User have to compute and fix all issues before hand, then leave the rest of the issues on the list.
Via [scripting](https://github.com/sn4k3/UVtools/tree/master/UVtools.ScriptSample) you can access issues with: `SlicerFile.IssueManager.GetIssues()` it hold an list of all [issues](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Layers/Issue.cs) ungrouped. Looping that object will return a `Issue` which can be a island, overhang, resin trap, suction cup, etc. You need to filter/order and do your math based on your criteria. It's a easy task but require some C# knowledge.

**Reply by jimb15122** [reply]
Thank you, sounds like the scripting is very powerful.  But when I used the term "overhang pixel", I simply meant any _pixel_ that is "on" in the current layer, but was "off" on the prior layer.  Most layers have these, and they should be easy to count, but they do not (necessarily) result in an "overhang" issue, unless some additional quantity criteria is met.
So to do what I want, I would need to need to following:
1. Ability to iterate through all layers, and compare each pixel in the layer with the prior layer.
2. count number of "on" pixels that were also "on" in the prior layer.
3. count number of "on" pixels that were "off" in the prior layer.
4. compute the ratio and compute the exposure time setting.
5. set a layer property with the adjusted time setting.

I am also not sure if any sla file formats would even support "layer level" time settings.  In my case, I'm using cxdlp.    

Your idea of using the overhang issues would also solve this.  All I really want to do is to be able to set a longer exposure time, for layers that have unresolved overhangs.  It seems this would be a nice feature to add to the product itself (a new overhang exposure time setting).  It would allow improved quality for those pesky high overhang layers, without having to increase the exposure time for all layers.

**Reply by jimb15122** [reply]
I noticed that the "repair layers and issues" dialog has no mention of "overhang" issues.  Perhaps the ability to specify an "additional exposure time for overhang layers" on this dialog would do the trick.  Does anyone else think this would be a viable way to "fix" overhangs??? Feature request??

**Reply by sn4k3** [reply]
> But when I used the term "overhang pixel", I simply meant any pixel that is "on" in the current layer, but was "off" on the prior layer.

That's simpler but then you are not taking into account supported masses that would print fine without adjustment.

count number of "on" pixels that were also "on" in the prior layer. **-- Is is two lines of code** 
count number of "on" pixels that were "off" in the prior layer.  **-- Is is two lines of code**
compute the ratio and compute the exposure time setting.  **-- Is is two line of code**
set a layer property with the adjusted time setting. **-- Is is one line of code**

With about 20 lines of code you can have your idea done, as with that criteria is very easy to implement, so totally possible to write.

> In my case, I'm using cxdlp

Not compatible with that format. No per layer adjustment.

> It seems this would be a nice feature to add to the product itself (a new overhang exposure time setting). It would allow improved quality for those pesky high overhang layers, without having to increase the exposure time for all layers.

I really don't know, as it will increase time per that layer and not just the region and different times can create visible lines.
I can invite people to discuss and vote for the feature. As i'm not a resin printer i don't know the goods or bads about this and require more people to join and discuss the ideia.

> I noticed that the "repair layers and issues" dialog has no mention of "overhang" issues.

Overhangs are supported masses with large unsupported areas between supports, must be fixed on slicer with more supports.

> Perhaps the ability to specify an "additional exposure time for overhang layers" on this dialog would do the trick.

Is best to run additional tool for that, as that is not the proper fix for overhangs

> Does anyone else think this would be a viable way to "fix" overhangs??? Feature request??

I will call more people into this.

**Reply by OEM-IS** [reply]
So as I mentioned on facebook, I had started looking into this issue kinda from the opposite direction asking on reddit and facebook for info - [Calculating the number of supports needed per cm^2 to overcome the peel force of a particular layer.](https://old.reddit.com/r/resinprinting/comments/w8xmcu/calculating_the_number_of_supports_needed_per_cm2/)  and on the UVTools Facebook here https://www.facebook.com/groups/uvtools/posts/726046651838174/?comment_id=726561345120038&reply_comment_id=727103168399189&notif_id=1658882644412021&notif_t=group_comment_mention  - nd tangentially when trying to find a way to calculate how to test or calculate lift height [The effects of lowering lift height and how to test whats sufficient?](https://old.reddit.com/r/resinprinting/comments/w69k8d/the_effects_of_lowering_lift_height_and_how_to/)

**Reply by OEM-IS** [reply]
I did find a couple of papers on the subject. Many of the topics go way over my head (my background is comp-sci/3d-modelling/ancient world studies :P) but they began to give me an idea of some of the forces involved and ways to test them, as well as findings from other tests.   

While trying to find links for the PDF's I had, I came across this page with a bunch of extra ones from some people doing what looks like a university research project in  Materials Engineering.  https://wiki.ubc.ca/Course:MTRL466-Supports4SLA 

Anyway here are some papers I found which pretty much all relate to support structures

Numerical and experimental investigation of effective parameters on separation force in bottom-up stereolithography process
- https://ajme.aut.ac.ir/article_4162.html


Perspective on 3D printing of separation membranes and comparison to related unconventional fabrication techniques
- sciencedirect.com/science/article/pii/S0376738816304215

Examining the Relationship between Forces During Stereolithography 3D Printing and Geometric Parameters of the Model
-https://www.researchgate.net/publication/292347157_Examining_the_Relationship_between_Forces_During_Stereolithography_3D_Printing_and_Geometric_Parameters_of_the_Model

Branching Support Structures for 3D Printing (paper)
- https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=2ahUKEwicp_-RzqndAhXBKH0KHQvDB3gQFjAAegQIBxAC&url=https%3A%2F%2Fd2f99xq7vri1nk.cloudfront.net%2FSupportStructuresSG14.pdf&usg=AOvVaw1ZJWFc7NlZ_eKXTptBUOmy

Efficient Support Structure Generation for Digital Fabrication (paper)
- https://www.cs.purdue.edu/cgvlab/www/publications/Vanek14SGP/


Computer Aided Design of Support Structures for Stereolithographic Components
- https://www.academia.edu/33636971/Computer_Aided_Design_of_Support_Structures_for_Stereolithographic_Components


Computer-aided build style decision support for stereolithography
- https://www.researchgate.net/publication/239397662_Computer-aided_build_style_decision_support_for_stereolithography

Support Structures for Additive Manufacturing:A Review
- https://www.mdpi.com/2504-4494/2/4/64/pdf

Research by NewPro3D

SUMMARY OF FINDINGS – OPTIMIZATION OF 3D PRINTED PART MECHANICAL PROPERTIES
- https://wiki.ubc.ca/images/5/55/SUMMARY_OF_FINDINGS_%E2%80%93_OPTIMIZATION_OF_3D_PRINTED_PART_MECHANICAL_PROPERTIES.pdf


Pull force in NewPro3D process
EFFECTS OF PRINT WINDOW MATERIAL ON SEPARATION FORCE OF 3D PRINTED OBJECTS IN DLP SLA PRINTING PROCESSES
- https://wiki.ubc.ca/images/f/f5/Pull_Force_Testing_-_Lab_Report_v4.pdf

EFFECTS OF PRINT AND POSTCURE PARAMETERS ON MECHANICAL PROPERTIES OF 3D PRINTED PARTS
- https://wiki.ubc.ca/images/f/ff/EFFECTS_OF_PRINT_AND_POST-CURE_PARAMETERS_ON_MECHANICAL_PROPERTIES_OF_3D_PRINTED_PARTS.pdf

**Reply by OEM-IS** [reply]
There was also a document with a summary of the forces involved from someone at NewPro3d

Factors that affect support structures:
• Resin viscosity – high viscosity resins require more supports to counteract vacuum forces as the
part is separated from the print window
• Green strength of the material – Materials with low stiffness in their green state require more
supports to resist deflection, this is compounded by the fact that support structures are made of
the same flexible material as the object being printed
• Part Geometry and size – vacuum forces increase exponentially with cross-sectional area.
Vacuum forces are also dependent on part geometry, a circular shape with the same cross
sectional area as a star shape will generate more vacuum forces as the resin must travel further
to reach the center of the object
• Angle of a part – more support structures are required as parts deviate from vertical geometries
to counteract gravity
• Platform ascent speed – increasing platform movement speed increases vacuum forces
• More support structures = more resin and more post processing! Optimizing support structures
will reduce costs, post processing requirements and improve product quality
• Would be beneficial for the students to design simple test models that will show us at which
angle support structures are required as well as models with increasing overhangs to determine
at which distance form a part support structures are required on overhangs
• Design considerations
o Test models should not have features to be measured within the first 2mm (vertical),
increased power is used in the first layers of printing that could skew results
o Smaller (within reason) test models are better as they will use less resin and produce
results more quickly
